runtime.md

```diff
@@ -1,5 +1,47 @@
 # Runtime and Lifecycle
 
+## State
+
+The runtime state for a container is persisted on disk so that external tools can consume and act on this information.
+The runtime state is stored in a JSON encoded file.
+It is recommended that this file is stored in a temporary filesystem so that it can be removed on a system reboot.
+On Linux based systems the state information should be stored in `/run/oci/containers`.
+The directory structure for a container is `/run/oci/containers/<containerID>/state.json`.
+By providing a default location that container state is stored external applications can find all containers running on a system.
+
+* **version** (string) Version of the OCI specification used when creating the container.
+* **id** (string) ID is the container's ID.
+* **pid** (int) Pid is the ID of the main process within the container.
+* **root** (string) Root is the path to the container's bundle directory.
+
+The ID is provided in the state because hooks will be executed with the state as the payload.
+This allows the hook to perform clean and teardown logic after the runtime destroys it's own state.
+The root directory to the bundle is provided in the state so that consumers can find the container's configuration and rootfs.
+
+*Example*
+
+```json
+{
+    "id": "oci-container",
+    "pid": 4422,
+    "root": "/containers/redis"
+}
+```
+
+Linux systems add some platform specific information to the state.
+
+* **externalFds** Paths to the container's open file descriptors.
+
+*Example Linux*
+
+```json
+{
+    "externalFds": [
+        "/proc/33/fd/1"
```

Comment:
> But maybe the important information in externalFds is just the
order and fact that a given file descriptor was passed from the
launching runtime

externelFds defines correspondence between descriptor numbers and file objects in a moment of start.

"externalFds": {"pipe:[18132]", "pipe:[18132]", "pipe:[18132]", "socket[123456]"}
This means that a process was executed with stdout:"pipe:[18132]", ..., 3:"socket[123456]"

> How does the restoring application know to create that socket?
The user knows how to start a process and how to prepare descriptors for it. If he wants to dump and restore the process, he must know how to create these descriptors in this case too.

> How do you know which of those should replace socket:[3197260]?

Lets imagine that we start a process with one extra descriptor:
```go
process := libcontainer.Process{
    Args: ...,
    Stdin: stdin,
    Stdout: stdout,
    Stderr: stderr,
    ExtraFiles: []*os.File{sk},
}
err = container.Start(&process)
```

Now we have:
  "externalFds": {
    "stdin": "pipe:[18132]",
    "stdout": "pipe:[18133]",
    "stderr": "pipe:[18134]",
    "socket": "socket:[3197260]"
  }

Then we make dump and restore this CT. On restore we need to set new descriptors for stdin, stdout, stderr ans sk.

```go
process := libcontainer.Process{
    Args: ...,
    Stdin: stdin_new,
    Stdout: stdout_new,
    Stderr: stderr_new,
    ExtraFiles: []*os.File{sk_new},
}
err = container.Restore(&process)
```
ExtraFiles doesn't include stdin, stdout, stderr, so the full list of descriptors looks like this:

Files: {stdin_new, stdout_new, stderr_new, sk_new}

"CRIU restore" will restore Files[i] instead of externalFds[i].